rust/xaynet-analytics/src/data_combination/data_points/was_active_each_past_period.rs

```diff
@@ -0,0 +1,58 @@
+use chrono::{DateTime, Utc};
+use std::collections::HashMap;
+
+use crate::{
+    data_combination::data_points::data_point::{
+        CalcWasActiveEachPastPeriod,
+        CalculateDataPoints,
+        DataPointMetadata,
+    },
+    data_provision::analytics_event::AnalyticsEvent,
+};
+
+impl CalcWasActiveEachPastPeriod {
+    pub fn new(
+        metadata: DataPointMetadata,
+        events: Vec<AnalyticsEvent>,
+        period_thresholds: Vec<DateTime<Utc>>,
+    ) -> CalcWasActiveEachPastPeriod {
+        CalcWasActiveEachPastPeriod {
+            metadata,
+            events,
+            period_thresholds,
+        }
+    }
+
+    // TODO: this could possibly later be optimised by sorting events by timestamp and caching the last timestamp added to the HashMap
+    fn group_timestamps_by_period_threshold(&self) -> HashMap<DateTime<Utc>, Vec<DateTime<Utc>>> {
+        let mut timestamps_by_period_threshold = HashMap::new();
+        for these_thresholds in self.period_thresholds.windows(2) {
+            let newer_threshold = these_thresholds.first().unwrap();
```

Comment:
Perhaps we can add a "safety comment" describing why it is safe to `unwrap` here

something like:
```rust
// safe unwrap: `windows` guarantees that there are at least two elements. 
// If `period_thresholds` contains less than two elements, this code block is not executed
```

Reply:
will do